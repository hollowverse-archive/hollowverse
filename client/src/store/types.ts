import { IAlgoliaSearchResults } from 'vendor/algolia';
import { HvError as AppError } from 'typings/typeDefinitions';
import { INotablePersonSchema, IUserSchema } from 'typings/dataSchema';

export type StoreState = {
  // Search
  searchInputValue: string;
  searchResults: IAlgoliaSearchResults | undefined;
  isSearchPending: boolean;
  lastSearchTerm: string;

  // Authentication
  loginStatus: facebookSdk.LoginStatus;
  isLoginPending: boolean;
  isLogoutPending: boolean;

  // General
  error: AppError;
  displayWarning: boolean;
  isNavMenuOpen: boolean;

  // Profile
  createProfileUrlInputValue: string;

  // Notable Person
  notablePerson: INotablePersonSchema | undefined;

  // User
  userData: IUserSchema | undefined;
};

export type StoreKey = keyof StoreState;

/** A map of all app actions to their corresponding payloads */
export type PayloadsByActionType = {
  // General
  setError: AppError;
  toggleWarning: boolean;

  // Search
  setSearchInputValue: string;
  requestSearchResults: string;
  setIsSearchPending: boolean;
  setSearchError: string;
  setSearchResults: IAlgoliaSearchResults | undefined;
  setLastSearchTerm: string;
  navigateToSearch: string;

  // Authentication
  requestLogin: void;
  setLoginStatus: facebookSdk.LoginStatus;
  requestUpdateLoginStatus: void;
  setIsLoginPending: boolean;
  setIsLogoutPending: boolean;
  requestLogout: void;

  // Navigation/Routing
  setIsNavMenuOpen: boolean;

  // Profile
  setCreateProfileUrlInputValue: string;

  // Notable person
  requestNotablePerson: string;
  setNotablePerson: INotablePersonSchema;

  // User
  setUserData: IUserSchema;
};

/**
 * This type covers all possible action types that may be dispatched
 * throught the app.
 */
export type ActionType = keyof PayloadsByActionType;

/**
 * A typed payload.
 * It must one of the payloads defined above,
 * and must correspond to the action type T.
 */
export type Payload<T extends ActionType> = PayloadsByActionType[T];

export type GenericPayload = Payload<ActionType>;

/**
 * Typed app-specific action
 * The `payload` must correspond to the `type`.
 * This provides improved type checking and prevents dispatching an action
 * with the wrong payload type.
 */
export type Action<T extends ActionType> = {
  type: T;
  payload: Payload<T>;
};

/** Generic action, used when we do not care about what the type */
export type GenericAction = {
  type: ActionType;
  payload?: GenericPayload;
};

/** A typed version of Redux store.dispatch() */
export type Dispatch<T extends ActionType> = (
  action: Action<T>,
) => typeof action;

/**
 * A Generic dispatch function.
 * Used when we do not care about the action type
 * but it still requiers that the type of action is one of
 * the app actions defined above.
 */
export type GenericDispatch = Dispatch<ActionType>;

export type ActionCreator<T extends ActionType> = (
  payload: Payload<T>,
) => Action<T>;

export type GenericActionCreator = ActionCreator<ActionType>;

export type Reducer<S, A extends ActionType> = (
  state: S,
  action: Action<A>,
) => S;

export type ReducerMap = {
  [Key in StoreKey]: Reducer<StoreState[Key], ActionType>
};

export type ActionToReducerMap<Key extends StoreKey> = {
  [A in ActionType]: Reducer<StoreState[Key], A>
};

/**
 * The default props passed by `react-redux`'s `connect()`
 * when no `mapDispatchProps` is passed as the second argument.
 * This type provides strong typing for `dispatch()` and it only accepts
 * actions defined in our app.
 */
export type DefaultDispatchProps = {
  dispatch<A extends GenericAction>(action: A): typeof action;
};
