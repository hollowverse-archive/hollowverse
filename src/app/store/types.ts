import {
  RouterState,
  LocationChangeAction,
  RouterAction,
} from 'react-router-redux';
import { AsyncResult } from 'helpers/asyncResults';
import { AlgoliaResponse } from 'algoliasearch';
import { NotablePersonQuery, ViewerQuery } from 'api/types';
import { DeepPartial } from 'typings/typeHelpers';

export type ResolvedData = {
  notablePersonQuery: NotablePersonQuery | null;
  searchResults: AlgoliaResponse | null;
  viewer: ViewerQuery | null;
};

export type ResolvedDataKey = keyof ResolvedData;

export type RequestDataPayload<
  Key extends ResolvedDataKey = ResolvedDataKey
> = {
  /**
   * Whether to keep the results of the previous request while loading
   * the new results
   */
  keepStaleData: boolean;

  /**
   * An incomplete optimistic version of the data that is expected
   * to be loaded.
   */
  optimisticResponse?: DeepPartial<ResolvedData[Key]>;

  /**
   * The key used to store the results in Redux state
   */
  key: ResolvedDataKey;

  /**
   * A unique identifier for the resolve request, if this changes,
   * `load()` will be called again
   */
  requestId: string | null;

  /**
   * (Optional)
   * The page path for which this data request is triggered, useful for logging
   */
  forPage?: string;

  /** An asynchronous function that fetches the data */
  load(): Promise<ResolvedData[Key]>;
};

export type SetResolvedDataPayload<
  Key extends ResolvedDataKey = ResolvedDataKey
> = {
  /**
   * The key used to store the results in Redux state
   */
  key: Key;

  /**
   * (Optional)
   * The page path for which this data request is triggered, useful for logging
   */
  forPage?: string;

  data: AsyncResult<ResolvedData[Key]> & {
    requestId: string | null;
  };
};

type SerializableError = {
  name: string;
  message: string;
  stack?: string;
};

/** A map of all app actions to their corresponding payloads */
export type ActionTypeToPayloadType = {
  GO_TO_SEARCH: void;
  SEARCH_QUERY_CHANGED: {
    query: string;
  };
  SET_SHOULD_FOCUS_SEARCH: boolean;
  SET_STATUS_CODE:
    | {
        code: 200 | 404;
      }
    | {
        code: 500;
        error?: SerializableError;
      }
    | {
        code: 301 | 302;
        redirectTo: string;
      };
  REQUEST_DATA: RequestDataPayload;
  SET_RESOLVED_DATA: SetResolvedDataPayload;
  UNHANDLED_ERROR_THROWN: SerializableError & {
    source?: string;
    line?: number;
    column?: number;
    location: Partial<Pick<Location, 'pathname' | 'search' | 'hash'>>;
  };
  PAGE_LOAD_FAILED: { path: string; error?: SerializableError };
  PAGE_LOAD_SUCCEEDED: { path: string };
  PAGE_REDIRECTED: {
    statusCode: 301 | 302;
    from: string;
    to: string;
  };
  /** Value is the path to the selected page */
  SEARCH_RESULT_SELECTED: string;
  SET_ALTERNATIVE_SEARCH_BOX_TEXT: string | null;
  '@@router/LOCATION_CHANGE': LocationChangeAction['payload'];
  '@@router/CALL_HISTORY_METHOD': RouterAction['payload'];
  FACEBOOK_AUTH_RESPONSE_CHANGED: facebookSdk.AuthResponse;
  TOGGLE_AUTH_STATUS: undefined;
};

export type AppState = {
  statusCode: 301 | 302 | 404 | 200 | 500;
  redirectionUrl: string | null;
  shouldFocusSearch: boolean;
  /**
   * Used to display text in place of the search box when the user
   * scrolls the page down, like the notable person's name on the notable
   * person's page.
   */
  alternativeSearchBoxText: string | null;
  resolvedData: {
    [K in keyof ResolvedData]: AsyncResult<ResolvedData[K] | null> & {
      requestId: string | null;
    }
  };
  facebookAuthResponse: facebookSdk.AuthResponse;
};

/**
 * Contains AppState as well as other state keys that are
 * required by external modules
 */
export type StoreState = Readonly<AppState & { routing: RouterState }>;

export type StoreKey = keyof AppState;

/**
 * This type covers all possible action types that may be dispatched
 * throughout the app.
 */
export type ActionType = keyof ActionTypeToPayloadType;

/**
 * A typed payload.
 * It must one of the payloads defined above,
 * and must correspond to the action type `T`.
 */
export type Payload<T extends ActionType> = ActionTypeToPayloadType[T];

export type GenericPayload = Payload<ActionType>;

/** Generic action, used when we do not care about what the type */
export type GenericAction = {
  type: ActionType;
  payload?: GenericPayload;
};

/**
 * Typed app-specific action
 * The `payload` must correspond to the `type`.
 * This provides improved type checking and prevents dispatching an action
 * with the wrong payload type.
 */
export type Action<T extends ActionType = ActionType> = {
  type: T;
  payload: Payload<T>;
};

export type ActionCreator<T extends ActionType> = (
  payload: Payload<T>,
) => Action<T>;

export type Reducer<S> = (state: S, action: Action) => S;

export type ReducerMap<State extends object = AppState> = {
  [Key in keyof State]: Reducer<State[Key]> | any
};

export type ActionToReducerMap<Key extends StoreKey> = {
  [A in ActionType]: Reducer<AppState[Key]>
};

export type LoggedAction = Action & {
  timestamp: Date;
};

export type LogBatch<A extends Action = LoggedAction> = {
  actions: A[];
  sessionId: string;

  /**
   * Note: user agent cannot be read from the User-Agent header, instead
   * it must be sent in the action payload. This is because the User-Agent
   * header is not whitelisted in CloudFront so it's stripped before the request
   * is seen by the log endpoint.
   * White-listing the User-Agent header is an extremely
   * bad idea because it means a cache object will be created for every
   * browser/browser version/device combination.
   */
  userAgent?: string;
};
