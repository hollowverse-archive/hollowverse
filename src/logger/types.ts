import { Action } from '../app/store/types';

export type LoggedAction = Action & {
  timestamp: string;
};
