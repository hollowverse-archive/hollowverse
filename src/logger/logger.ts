// tslint:disable no-console
import { DateTime } from 'luxon';
import bluebird from 'bluebird';
import { SourceMapConsumer } from 'source-map';
import got from 'got';
import { URL } from 'url';
import flattenObject from 'flat';

import { LoggedAction } from './types';
import { isActionOfType } from '../app/store/helpers';

const { BRANCH, COMMIT_ID, SUMO_COLLECTOR_ID } = process.env;

const RECEIVER_URL =
  'https://endpoint2.collection.us2.sumologic.com/receiver/v1/http/';

// tslint:disable-next-line no-non-null-assertion
const COLLECTOR_URL = new URL(SUMO_COLLECTOR_ID!, RECEIVER_URL).toString();

const transformActionForLogging = async (
  action: LoggedAction,
): Promise<LoggedAction> => {
  if (isActionOfType(action, 'UNHANDLED_ERROR_THROWN')) {
    const { message, source, line, column } = action.payload;
    if (source && line && column) {
      const { body: sourceMap } = await got(`${source}.map?branch=${BRANCH}`, {
        json: true,
      });
      const consumer = new SourceMapConsumer(sourceMap);
      const originalPosition = consumer.originalPositionFor({ line, column });

      return {
        ...action,
        payload: {
          message,
          ...originalPosition,
        },
      };
    }
  }

  return action;
};

export const convertObjectsToLines = (
  additionalProps?: Record<string, any>,
) => (action: LoggedAction) => {
  const { timestamp, type, ...rest } = action;

  const normalizedDate = DateTime.fromISO(timestamp, { zone: 'UTC' });

  const pairs = Object.entries(
    flattenObject({
      type,
      ...additionalProps,
      ...rest,
    }),
  ).map(([key, value]) => `[${key}=${value}]`);

  return `${normalizedDate} ${pairs.join(' ')}`;
};

export async function log(actions: LoggedAction[]) {
  const lines = await bluebird
    .map(actions, transformActionForLogging)
    .map(convertObjectsToLines({ branch: BRANCH, commit: COMMIT_ID }));

  await got.post(COLLECTOR_URL, {
    body: lines.join('\n'),
    headers: {
      'Content-Type': 'text/plain',
      'X-Sumo-Category': 'hollowverse.com',
      'X-Sumo-Host': 'Lambda',
    },
  });
}
