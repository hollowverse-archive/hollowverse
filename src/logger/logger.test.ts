import { convertObjectsToLines } from './logger';
import { LoggedAction } from './types';

describe('Logger', () => {
  it('should work', () => {
    const loggedActions: LoggedAction[] = [
      {
        type: 'UNHANDLED_ERROR_THROWN',
        payload: {
          message: 'Unknown error',
          column: 0,
          line: 0,
        },
        timestamp: '2018-05-05T18:51:07.908Z',
      },
      {
        type: 'PAGE_LOAD_FAILED',
        payload: '/Tom_Hanks',
        timestamp: '2018-05-05T19:18:49.053Z',
      },
      {
        type: 'PAGE_LOAD_SUCCEEDED',
        payload: '/Tom_Hanks',
        timestamp: '2018-05-05T19:18:49.053Z',
      },
    ];

    for (const action of loggedActions) {
      expect(
        convertObjectsToLines({ branch: 'master', commit: '123456' })(action),
      ).toMatchSnapshot();
    }
  });
});
