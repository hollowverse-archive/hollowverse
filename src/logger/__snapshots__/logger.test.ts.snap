// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should work 1`] = `"2018-05-05T18:51:07.908Z [type=UNHANDLED_ERROR_THROWN] [branch=master] [commit=123456] [payload.message=Unknown error] [payload.column=0] [payload.line=0]"`;

exports[`Logger should work 2`] = `"2018-05-05T19:18:49.053Z [type=PAGE_LOAD_FAILED] [branch=master] [commit=123456] [payload=/Tom_Hanks]"`;

exports[`Logger should work 3`] = `"2018-05-05T19:18:49.053Z [type=PAGE_LOAD_SUCCEEDED] [branch=master] [commit=123456] [payload=/Tom_Hanks]"`;
